sky/provision/oci/query_utils.py

```diff
@@ -468,5 +486,185 @@ def delete_vcn(cls, net_client, skypilot_vcn, skypilot_subnet,
+        nsg_id = cls.find_nsg(region, nsg_name, True)
+
+        filters = {constants.TAG_RAY_CLUSTER_NAME: cluster_name}
+        insts = query_helper.query_instances_by_tags(filters, region)
+        for inst in insts:
+            vnic = cls.get_instance_primary_vnic(
+                region=region,
+                inst_info={
+                    'inst_id': inst.identifier,
+                    'ad': inst.availability_domain,
+                    'compartment': inst.compartment_id,
+                })
+            nsg_ids = vnic.nsg_ids
+            if not nsg_ids:
+                net_client.update_vnic(
+                    vnic_id=vnic.id,
+                    update_vnic_details=oci_adaptor.oci.core.models.
+                    UpdateVnicDetails(nsg_ids=[nsg_id],
+                                      skip_source_dest_check=False),
+                )
+
+        # pylint: disable=line-too-long
+        list_nsg_rules_resp = net_client.list_network_security_group_security_rules(
+            network_security_group_id=nsg_id,
+            direction='INGRESS',
+            sort_by='TIMECREATED',
+            sort_order='DESC',
+        )
+
+        ingress_rules: List = list_nsg_rules_resp.data
+        existing_port_ranges: List[str] = []
+        for r in ingress_rules:
+            if r.tcp_options:
+                rule_port_range = f'{r.tcp_options.destination_port_range.min}-{r.tcp_options.destination_port_range.max}'
+                existing_port_ranges.append(rule_port_range)
+
+        new_ports = resources_utils.port_ranges_to_set(ports)
+        existing_ports = resources_utils.port_ranges_to_set(
+            existing_port_ranges)
+        if new_ports.issubset(existing_ports):
+            # ports already contains in the existing rules, nothing to add.
+            return
+
+        union_ports = new_ports.union(existing_ports)
+        union_port_ranges = resources_utils.port_set_to_ranges(union_ports)
+
+        new_rules = [
+            oci_adaptor.oci.core.models.AddSecurityRuleDetails(
+                direction='INGRESS',
+                protocol='6',
+                is_stateless=False,
+                source=oci_utils.oci_config.VCN_CIDR_INTERNET,
+                source_type='CIDR_BLOCK',
+                tcp_options=oci_adaptor.oci.core.models.TcpOptions(
+                    destination_port_range=oci_adaptor.oci.core.models.
+                    PortRange(min=cls.get_range_min_max(range)[0],
+                              max=cls.get_range_min_max(range)[1]),),
+                description=oci_utils.oci_config.SERVICE_PORT_RULE_TAG,
+            ) for range in union_port_ranges
+        ]
+
+        net_client.add_network_security_group_security_rules(
+            network_security_group_id=nsg_id,
+            add_network_security_group_security_rules_details=oci_adaptor.oci.
+            core.models.AddNetworkSecurityGroupSecurityRulesDetails(
+                security_rules=new_rules),
+        )
```

Comment:
Will the new rules have overlap with the old rules? Should we only create new ports?

Reference:
https://github.com/skypilot-org/skypilot/blob/1c04aefbd42d9e59333f2757f066e5fad32b45ec/sky/provision/aws/instance.py#L804-L805

Reply:
In addition to the above comments, my confusion is that seems like here we are **adding new rules**, instead of **overriding all rules in nsg**, so probably we should use subtraction? If here is overriding all rules in nsg, then the union looks good to me.